azure-functions-powershell-library/WorkerIndexingHelper.cs

```diff
@@ -0,0 +1,197 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using AzureFunctions.PowerShell.SDK;
+using System.Collections.ObjectModel;
+using System.Management.Automation.Language;
+
+namespace Microsoft.Azure.Functions.PowerShellWorker
+{
+    internal class WorkerIndexingHelper
+    {
+        internal static List<FunctionInformation> IndexFunctions(string baseDir)
+        {
+            if (!Directory.Exists(baseDir))
+            {
+                throw new FileNotFoundException();
+            }
+            List<FileInfo> powerShellFiles = GetPowerShellFiles(Directory.CreateDirectory(baseDir));
+            List<FunctionInformation> rpcFunctionMetadatas = new List<FunctionInformation>();
+
+            foreach (FileInfo powerShellFile in powerShellFiles)
+            {
+                rpcFunctionMetadatas.AddRange(IndexFunctionsInFile(powerShellFile));
+            }
+
+            return rpcFunctionMetadatas;
+        }
+
+        private static IEnumerable<FunctionInformation> IndexFunctionsInFile(FileInfo powerShellFile)
+        {
+            List<FunctionInformation> fileFunctions = new List<FunctionInformation>();
+            var fileAst = Parser.ParseFile(powerShellFile.FullName, out _, out ParseError[] errors);
+            if (errors.Any())
+            {
+                throw new Exception($"Couldn't parse the file: {powerShellFile.FullName}");
+                // TODO: Probably don't throw here?
+                //return fileFunctions;
+            }
+            if (powerShellFile.Extension == ".ps1") 
+            {
+                // parse only the file param block, return one RpcFunctionMetadata assuming the file is the entry point
+                var paramAsts = fileAst.ParamBlock;
+                if (paramAsts != null && paramAsts.Attributes.Where(x => x.TypeName.ToString() == "Function").Any()) 
+                {
+                    // This is a function, return it 
+                    fileFunctions.Add(CreateRpcMetadataFromFile(powerShellFile.FullName));
+                }
+            }
+            else if (powerShellFile.Extension == ".psm1")
+            {
+                var potentialFunctions = fileAst.FindAll(x => x is FunctionDefinitionAst, false);
+                foreach (var potentialFunction in potentialFunctions)
+                {
+                    var matchingBlocks = potentialFunction.FindAll(x => x is ParamBlockAst && ((ParamBlockAst)x).Attributes.Where(z => z.TypeName.ToString() == "Function").Any(), true);
+                    if (matchingBlocks.Any()) {
+                        //This function is one we need to register
+                        fileFunctions.Add(CreateRpcMetadataFromFunctionAst(powerShellFile.FullName, (FunctionDefinitionAst)potentialFunction));
+                    }
+                }
+                // parse all function definitions, return as many RpcFunctionMetadatas as exist in the file
```

Comment:
there seems to be a missing `else` case in here. Should we add it and add a warning of some kind?

Reply:
No, there may be any number of functions defined in a .psm1 file that the end user does not want to register as an Azure Function. We'll just skip those function definitions and move on

Reply:
Thanks for the clarification, @andystaples. In that case, my recommendation would be to add a comment here specifying precisely what you just described :-)

Reply:
Done, it'll come in whenever I do the next commit